sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -1192,51 +1198,326 @@ ProgramManager::getSYCLDeviceImages(const context &Ctx,
+      InputImpl->get_bin_image_ref(), InputImpl->get_context(), Devs,
+      bundle_state::object, InputImpl->get_kernel_ids_ref(), Prog);
+
+  std::vector<pi_device> PIDevices;
+  PIDevices.reserve(Devs.size());
+  for (const device &Dev : Devs)
+    PIDevices.push_back(getSyclObjImpl(Dev)->getHandleRef());
+
+  // TODO: Set spec constatns here.
+
+  // TODO: Handle zero sized Device list.
+  Plugin.call<PiApiKind::piProgramCompile>(
+      ObjectImpl->get_program_ref(), /*num devices=*/Devs.size(),
+      PIDevices.data(),
+      /*options=*/nullptr,
+      /*num_input_headers=*/0, /*input_headers=*/nullptr,
+      /*header_include_names=*/nullptr,
+      /*pfn_notify=*/nullptr, /*user_data*/ nullptr);
+
+  return createSyclObjFromImpl<device_image_plain>(ObjectImpl);
+}
+
+std::vector<device_image_plain>
+ProgramManager::link(const std::vector<device_image_plain> &DeviceImages,
+                     const std::vector<device> &Devs,
+                     const property_list &PropList) {
+  (void)PropList;
+
+  std::vector<pi_program> PIPrograms;
+  PIPrograms.reserve(DeviceImages.size());
+  for (const device_image_plain &DeviceImage : DeviceImages)
+    PIPrograms.push_back(getSyclObjImpl(DeviceImage)->get_program_ref());
+
+  std::vector<pi_device> PIDevices;
+  PIDevices.reserve(Devs.size());
+  for (const device &Dev : Devs)
+    PIDevices.push_back(getSyclObjImpl(Dev)->getHandleRef());
+
+  const context &Context = getSyclObjImpl(DeviceImages[0])->get_context();
+
+  const detail::plugin &Plugin = getSyclObjImpl(Context)->getPlugin();
+
+  RT::PiProgram LinkedProg = nullptr;
+  RT::PiResult Error = Plugin.call_nocheck<PiApiKind::piProgramLink>(
+      getSyclObjImpl(Context)->getHandleRef(), PIDevices.size(),
+      PIDevices.data(),
+      /*options=*/nullptr, PIPrograms.size(), PIPrograms.data(),
+      /*pfn_notify=*/nullptr,
+      /*user_data=*/nullptr, &LinkedProg);
+
+  (void)Error;
+  // TODO: Add error handling
+
+  std::vector<kernel_id> KernelIDs;
+  for (const device_image_plain &DeviceImage : DeviceImages) {
+    // Duplicates are not expected here, otherwise piProgramLink should fail
+    KernelIDs.insert(KernelIDs.end(),
+                     getSyclObjImpl(DeviceImage)->get_kernel_ids().begin(),
+                     getSyclObjImpl(DeviceImage)->get_kernel_ids().end());
   }
+  // device_image_impl expects kernel ids to be sorted for fast search
+  std::sort(KernelIDs.begin(), KernelIDs.end(), LessByNameComp{});
 
-  return SYCLDeviceImages;
+  DeviceImageImplPtr ExecutableImpl =
+      std::make_shared<detail::device_image_impl>(
+          /*BinImage=*/nullptr, Context, Devs, bundle_state::object,
+          std::move(KernelIDs), LinkedProg);
+
+  // TODO: Make multiple sets of device images organized by devices they are
+  // compiled for.
+  return {createSyclObjFromImpl<device_image_plain>(ExecutableImpl)};
+}
+
+device_image_plain ProgramManager::build(const device_image_plain &DeviceImage,
```

Comment:
this is a big piece of highly complex code, most of it seems to be copy-paste from existing program-based build interface. could you please comment what is changed compared to the existing impl for faster review (or better mark with comments right here)?

Reply:
Added a comment which outlines main differences. 
Adding it here as well:
The function duplicates most of the code from existing getBuiltPIProgram.
The differences are:
 Different API - uses different objects to extract required info
 Supports caching of a program built for multiple devices

Reply:
thanks!